### DISCUSSION TITLE
Local Storage for session keys

### DISCUSSION THREAD

**Discussion by BramDevlaminck** [SELab-2/OSOC-2#390] [General]
At this moment each browser has 1 session key that is stored in the [`localStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage). The problem with this is that when we are using websockets and you are using multiple tabs (in the same browser). All these tabs are "fighting" for a session key and this creates conflicts (the keys will be used multiple times or an invalid key is used).

The solution seems to use [`sessionStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage). This way each tab would have his own session key.
The problem with this is that the key is lost when the tab is closed and this would make the auto sign-in impossible.

Has anyone any idea how we could fix this?

**Comment by Mouwrice** [comment]
We might have to make a separation of keys.
1. We have a `loginKey` that we store in persistent [`localStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) that is valid for multiple days to stay logged in in your browser.
2. We store the current `sessionKey` in [`sessionStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage) so we can make multiple requests per tab. This key should then only have a lifetime of say 5 minutes. Because of the short lifespan we no longer need to remove the key after it's request has been handled so that we can perform multiple requests at the same time using the same key.

Yes it would require some work, but it seems like there is no way going around it.

**Reply by BramDevlaminck** [reply]
This indeed seems the only way to achieve this. And it's also the only thing I could think of.

I guess when we start implementing this we should first try to only use the sessionStorage just to check if this indeed solves the issue with the keys when using websockets. When this works we can use the other key too to achieve automatic login.

**Reply by BramDevlaminck** [reply]
I've done a quick test by just only using the sessionStorage and this indeed works! So we should rewrite this with both kind of keys!

Replacing the sessionKeys from `SessionStorage` works without problem so I would to this when using a normal call to the backend.
Replacing the sessionKeys from `localStorage` is only needed upon login when we just have to check if the user is already logged in.
I'm just not yet sure how we can determine in the backend what kind of key you are receiving. This is needed if we want to use a different timeout for each kind.

**Reply by HuanYu-Tan** [reply]
what happens if the sessionKey expires?

**Reply by BramDevlaminck** [reply]
Good question. 5 minutes might be too tight. I guess that 3 hours (like Ufora does) is reasonable?

I'm not sure tho wat exactly will happen when the key expires. My guess is there will be an 'unauthorized' error

**Reply by Mouwrice** [reply]
If the sessionkey expires we could use the loginkey to request a new one i suppose

**Reply by BramDevlaminck** [reply]
won't this give the exact same problem when you don't touch a bunch of tabs for quite some time and then they all need a new key using that sessionKey from the `localStorage`?

**Reply by HuanYu-Tan** [reply]
I have another question in regards to the loginKey will it refresh when you make a request or is it only valid till x?

**Reply by BramDevlaminck** [reply]
I think this would still update? This has not given any problems so far, but we should think about this because this could give problems when using a really slow internet connection. In that case I think it would be possible to send 2 requests with that same key.

I don't know if this is "bad" because the login will just fail, you can try again immediately and this should fix itself once you are finally logged in with the first tab. Or am I wrong thinking this?

**Comment by BramDevlaminck** [comment]
We will use another way to solve this than what is described here, see #402
